File: app/src/main/java/com/kidozh/discuzhub/viewModels/AddBBSViewModel.java
package com.kidozh.discuzhub.viewModels;

import android.app.Activity;
import android.app.Application;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.MutableLiveData;

import com.kidozh.discuzhub.R;
import com.kidozh.discuzhub.activities.AddIntroActivity;
import com.kidozh.discuzhub.entities.bbsInformation;
import com.kidozh.discuzhub.results.AddCheckResult;
import com.kidozh.discuzhub.utilities.NetworkUtils;
import com.kidozh.discuzhub.utilities.URLUtils;
import com.kidozh.discuzhub.utilities.bbsParseUtils;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.net.URL;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class AddBBSViewModel extends AndroidViewModel {
    private final static String TAG = AddBBSViewModel.class.getSimpleName();

    public MutableLiveData<String> currentURLLiveData, errorTextLiveData;
    public @NonNull MutableLiveData<Boolean>  isLoadingLiveData, autoVerifyURLLiveData;
    public MutableLiveData<bbsInformation> verifiedBBS = new MutableLiveData<>(null);

    public AddBBSViewModel(@NonNull Application application) {
        super(application);
        currentURLLiveData = new MutableLiveData<>("");
        autoVerifyURLLiveData = new MutableLiveData<>(true);
        isLoadingLiveData = new MutableLiveData<>(false);
        errorTextLiveData = new MutableLiveData<>("");
    }

    public void verifyURL(){
        isLoadingLiveData.postValue(true);
        String base_url = currentURLLiveData.getValue();
        URLUtils.setBaseUrl(base_url);
        String query_url = URLUtils.getBBSForumInformationUrl();
        OkHttpClient client = NetworkUtils.getPreferredClient(getApplication());
        Request request;
        try{
            URL url = new URL(query_url);
//            if(url.getAuthority()!=null && url.getHost()!=null){
//                throw new Exception();
//            }
            request = new Request.Builder().url(query_url).build();
        }
        catch (Exception e){
            isLoadingLiveData.postValue(false);
            errorTextLiveData.postValue(getApplication().getString(R.string.bbs_base_url_invalid));
            e.printStackTrace();
            return;
        }

        Call call = client.newCall(request);
        Log.d(TAG,"Query check URL "+query_url);
        call.enqueue(new Callback() {

            @Override
            public void onFailure(Call call, IOException e) {
                isLoadingLiveData.postValue(false);
                e.printStackTrace();
                errorTextLiveData.postValue(getApplication().getString(R.string.network_failed));

            }

            @Override
            public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {
                isLoadingLiveData.postValue(false);
                String s;
                if(response.isSuccessful() && response.body()!=null){
                    s = response.body().string();
                    Log.d(TAG,"check response " +s);
                    AddCheckResult checkResult = bbsParseUtils.parseCheckInfoResult(s);
                    if(checkResult!=null){
                        bbsInformation bbsInfo = checkResult.toBBSInformation(base_url);
                        verifiedBBS.postValue(bbsInfo);
                        //new AddIntroActivity.addNewForumInformationTask(bbsInfo,activity).execute();
                    }
                    else {
                        errorTextLiveData.postValue(getApplication().getString(R.string.parse_failed));
                    }
                }
                else {
                    errorTextLiveData.postValue(getApplication().getString(R.string.parse_failed));
                }
            }
        });
    }
}

File: app/src/main/java/com/kidozh/discuzhub/viewModels/AddBBSViewModel.kt
package com.kidozh.discuzhub.viewModels

import android.app.Application
import android.util.Log
import android.util.Patterns
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.MutableLiveData
import com.kidozh.discuzhub.R
import com.kidozh.discuzhub.entities.Discuz
import com.kidozh.discuzhub.entities.ErrorMessage
import com.kidozh.discuzhub.results.AddCheckResult
import com.kidozh.discuzhub.services.DiscuzApiService
import com.kidozh.discuzhub.utilities.NetworkUtils
import com.kidozh.discuzhub.utilities.URLUtils
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response


class AddBBSViewModel(application: Application) : AndroidViewModel(application) {
    @JvmField
    var currentURLLiveData: MutableLiveData<String> = MutableLiveData("")

    @JvmField
    var errorTextLiveData: MutableLiveData<String> = MutableLiveData("")
    var errorMessageMutableLiveData: MutableLiveData<ErrorMessage?> = MutableLiveData(null)

    @JvmField
    var isLoadingLiveData: MutableLiveData<Boolean> = MutableLiveData(false)

    @JvmField
    var autoVerifyURLLiveData: MutableLiveData<Boolean> = MutableLiveData(true)

    @JvmField
    var verifiedBBS = MutableLiveData<Discuz?>(null)
    fun verifyURL() {

        var base_url = currentURLLiveData.value ?: return

        if(Patterns.WEB_URL.matcher(base_url).matches()){
            if(!base_url.startsWith("http://") && !base_url.startsWith("https://") ){
                base_url = "https://${base_url}"
            }
            isLoadingLiveData.postValue(true)
            URLUtils.setBaseUrl(base_url)

            val client = NetworkUtils.getPreferredClient(getApplication())
            val retrofit = NetworkUtils.getRetrofitInstance(base_url,client)
            val service = retrofit.create(DiscuzApiService::class.java)
            val checkCall = service.checkResult
            Log.d(TAG,"Get base url ${base_url} , ${checkCall.request().url}")
            checkCall.enqueue(object : Callback<AddCheckResult> {
                override fun onResponse(call: Call<AddCheckResult>, response: Response<AddCheckResult>) {
                    if (response.isSuccessful && response.body() != null) {
                        val checkResult = response.body() as AddCheckResult
                        if(!checkResult.siteName.equals("")){
                            verifiedBBS.postValue(checkResult.toBBSInformation(base_url))
                        }
                        else{
                            errorMessageMutableLiveData.postValue(ErrorMessage(getApplication<Application>().getString(R.string.check_result_incorrect_key),
                                getApplication<Application>().getString(R.string.check_result_incorrect_description))
                            )
                        }

                    }
                    else{
                        errorMessageMutableLiveData.postValue(ErrorMessage(response.code().toString(),
                            getApplication<Application>().getString(R.string.discuz_network_unsuccessful, response.message())))
                    }

                    isLoadingLiveData.postValue(false)
                }

                override fun onFailure(call: Call<AddCheckResult>, t: Throwable) {
                    errorMessageMutableLiveData.postValue(ErrorMessage(
                        getApplication<Application>().getString(R.string.discuz_network_failure_template),
                        if (t.localizedMessage == null) t.toString() else t.localizedMessage
                    ))
                    isLoadingLiveData.postValue(false)
                }

            })
        }
        else{
            errorTextLiveData.postValue(getApplication<Application>().getString(R.string.bbs_base_url_invalid))
        }





    }

    companion object {
        private val TAG = AddBBSViewModel::class.java.simpleName
    }

}